be/src/util/buffered_stream.cpp

```diff
@@ -64,4 +66,87 @@ Status BufferedInputStream::_read_data() {
     return Status::OK();
 }
 
+Status DefaultBufferedInputStream::get_bytes(const uint8_t** buffer, size_t offset, size_t* nbytes) {
+    seek_to(offset);
+    return get_bytes(buffer, nbytes, false);
+}
+
+// ===================================================================================
+
+SharedBufferedInputStream::SharedBufferedInputStream(RandomAccessFile* file) : _file(file) {}
+
+Status SharedBufferedInputStream::set_io_ranges(const std::vector<IORange>& ranges) {
+    if (ranges.size() == 0) {
+        return Status::OK();
+    }
+    std::vector<IORange> check(ranges);
+    std::sort(check.begin(), check.end());
+    // check io range is not overlapped.
+    for (size_t i = 1; i < ranges.size(); i++) {
+        if (check[i].offset < (check[i - 1].offset + check[i - 1].size)) {
+            return Status::RuntimeError("io ranges are overalpped");
+        }
+    }
+
+    std::vector<IORange> small_ranges;
+    for (const IORange& r : check) {
+        if (r.size > _options.max_buffer_size) {
+            SharedBuffer sb = SharedBuffer{.offset = r.offset, .size = r.size, .ref_count = 1};
+            _map.insert(std::make_pair(r.offset + r.size, sb));
+        } else {
+            small_ranges.emplace_back(r);
+        }
+    }
+
+    if (small_ranges.size() > 0) {
+        auto update_map = [&](size_t from, size_t to) {
+            // merge from [unmerge, i-1]
+            int64_t ref_count = (to - from + 1);
+            int64_t end = (small_ranges[to].offset + small_ranges[to].size);
+            SharedBuffer sb = SharedBuffer{.offset = small_ranges[from].offset,
+                                           .size = end - small_ranges[from].offset,
+                                           .ref_count = ref_count};
+            _map.insert(std::make_pair(sb.offset + sb.size, sb));
+        };
+
+        size_t unmerge = 0;
+        for (size_t i = 1; i < small_ranges.size(); i++) {
+            const auto& prev = small_ranges[i - 1];
+            const auto& now = small_ranges[i];
+            size_t now_end = now.offset + now.size;
+            size_t prev_end = prev.offset + prev.size;
+            if (((now_end - small_ranges[unmerge].offset) <= _options.max_read_size) &&
+                (now.offset - prev_end) <= _options.max_dist_size) {
+                continue;
+            } else {
+                update_map(unmerge, i - 1);
+                unmerge = i;
+            }
+        }
+        update_map(unmerge, small_ranges.size() - 1);
```

Comment:
should we check if unmerge smaller than small_ranges.size() - 1?

Reply:
No, you can tell from the code the invariant: `0 <= unmerge < small_ranges.size()`.

Reply:
```cpp
if (small_ranges.size() > 0) {
       // ...... 
        size_t unmerge = 0; 
        for (size_t i = 1; i < small_ranges.size(); i++) {
           // ......
            if (((now_end - small_ranges[unmerge].offset) <= _options.max_read_size) &&
                (now.offset - prev_end) <= _options.max_dist_size) {
                continue;
            } else {
                unmerge = i; // unmerge < small_ranges.size()
            }
        }
        update_map(unmerge, small_ranges.size() - 1);
}
```